lms/services/grouping.py

```diff
@@ -10,7 +10,7 @@
 class GroupingService:
     def __init__(self, db, application_instance_service):
         self._db = db
-        self._application_instance = application_instance_service.get_current()
+        self.application_instance = application_instance_service.get_current()
```

Comment:
I made `GroupingService.application_instance` a public attribute so that the unit tests can access it. A hack, perhaps? But I think it's acceptable: there's no real reason for this attribute to be public (only the tests access it). But there's no really strong reason why it needs to be private either. With `GroupingService.application_instance` public if anyone should want to know which `application_instance` `GroupingService` is going to use for any new groupings it creates, now they can. `GroupingService.application_instance` is arguably part of the "public interface" anyway even if nothing actually access it: it's the `application_instance` that it's going to use for new groupings that it creates for you. Any `parent` groupings that you pass to `GroupingService.upsert_with_parent()` must have `GroupingService.application_instance` as their `application_instance` or you'll get an `IntegrityError`

Reply:
I think "there's no real reason for this attribute to be public" might be the reason for it to be private we're looking for. Our tests and the unit under test are always tightly bound, so I wouldn't worry about accessing a private variable in the unit tests for an item. If it's expedient I'd just do it.

On the other hand, I don't actually have a problem with in this instance. I don't think any other code is going to start accessing the application instance via this service, and even if it did it's not hard to change.

Reply:
Duh, just noticed this is an instance not the service, but I think it's still fine.

To be clear, I don't think this would be fine on a model object, or something else we'd expect to be passed about a lot and consulted, as all attributes could be part of the interface, but our services are typically instantiated when you need them and then have a few methods called on them.

Reply:
> Our tests and the unit under test are always tightly bound, so I wouldn't worry about accessing a private variable in the unit tests for an item.

I guess I prefer that the unit tests work only through the public interface of the unit under test so that I can freely refactor the unit under test and as long as I haven't changed its public contract the unit tests should still pass.

That said I think you're right about the unit tests being very close to the unit under test, such that it's acceptable for them to access a private attribute if pragmatism demands it.

In this case I felt that `GroupingService.application_instance` _was_ arguably actually part of `GroupingService`'s public interface: it's the `application_instance` that `GroupingService` will use for any groupings that it creates for you when you call methods like `GroupingService.upsert()` etc. It doesn't actually _have_ to be public in the sense that no production code outside of `GroupingService` actually reads `GroupingService.application_instance` currently. But I think you can make some sort of argument that semantically it is part of the public interface.

You can also make a pretty strong argument that it doesn't matter a whole lot.

Anyway I left it public